server/src/test/java/io/spine/examples/chatspn/server/e2e/InteractionWithMessagesInPersonalChatTest.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2023, TeamDev. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Redistribution and use in source and/or binary forms, with or without
+ * modification, must retain the above copyright notice and the following
+ * disclaimer.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.spine.examples.chatspn.server.e2e;
+
+import io.spine.examples.chatspn.ChatId;
+import io.spine.examples.chatspn.account.UserChats;
+import io.spine.examples.chatspn.account.UserProfile;
+import io.spine.examples.chatspn.chat.ChatPreview;
+import io.spine.examples.chatspn.message.MessageView;
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import static io.spine.examples.chatspn.server.ExpectedOnlyAssertions.assertExpectedFieldsEqual;
+import static io.spine.examples.chatspn.server.e2e.given.PersonalChatTestEnv.chatPreview;
+import static io.spine.examples.chatspn.server.e2e.given.PersonalChatTestEnv.userChats;
+
+final class InteractionWithMessagesInPersonalChatTest extends ServerRunningTest {
+
+    @Test
+    @DisplayName("Users should be able to create account, personal chat, send messages, " +
+            "edit messages, remove messages, and delete a chat.")
+    void messageInteractionInPersonalChat() throws InterruptedException {
+        // Vlad and Artem passes registration.
+        TestUser vlad = new TestUser();
+        TestUser artem = new TestUser();
+
+        // Vlad finds Artem, creates a personal chat, and sees the chat's preview in `UserChats`.
+        UserProfile artemProfile = vlad.findUserBy(artem.email());
+        assertExpectedFieldsEqual(artemProfile, artem.profile());
+        ChatPreview expectedChat = vlad.createPersonalChatWith(artemProfile.getId());
+        ChatPreview vladChatView = vlad.readChats()
```

Comment:
The idea of such tests is to use the same approach as the desktop client app will.

In scope of it, we aren't going to be constantly polling the server for new chats, or new messages, or something like  that. General idea is as follows:

1. User navigates to some screen.
2. Subscriptions are created for each view that should be displayed on the screen.
3. For a single time, we explicitly query the view data from the server — just to have something to render on UI.
4. All subsequent UI changes are happening in one of two ways:
    * either we display some kind of input form to allow user send a command,
    * or we receive the updates from the server through the subscription updates, and then react accordingly (e.g., update the UI model).

Your current approach is poll-based, meaning you expect some things to appear on the read-side, and you know _exactly_ when to query them. This is not how it happens in the real-world scenario.

For instance, `vlad.findUserBy(email)` seems to be a proper thing to do, since that's a real person would explicitly do. But then explicit querying for chat previews is not — because the client code should have subscribed to the updates of this type of projections by now. Same goes for messages.

Now, to the expected values etc. I would create something like `ConversationScript`, which would hold the phrases users speak with to each other. Internally, that's something like `LinkedHashSet<UserId, YourMessageType>`, with the flag telling if the message was sent. Then at any given point of time you will know exactly which content is expected by each member of the conversation.

The problem with these ` expectedMessages.add(vlad.sendMessageTo(vladChatView.getId()));` is that the important part (message sending) is swallowed by the purely technical thing (preserving messages for comparison).

Please think this over, and let's later discuss vocally.

@Artem-Semenov-dev FYI.